include/onnxruntime/core/session/onnxruntime_c_api.h

```diff
@@ -3023,7 +3024,12 @@ struct OrtApi {
   */
   ORT_API2_STATUS(GetSparseTensorIndices, _In_ const OrtValue* ort_value, enum OrtSparseIndicesFormat indices_format, _Out_ size_t* num_indices, _Outptr_ const void** indices);
 
-  /// @}
+  /** \brief Returns the DeviceType of Tensor
+  * \param[in] ort_value ::OrtValue containing tensor.
+  * \param[out] int8_t :: OrtDeviceType Enum 
+  */
+  ORT_API2_STATUS(GetTensorDeviceType, _In_ const OrtValue* value, _Out_ int8_t* out);
```

Comment:
We should be using OrtMemoryInfo for this, using the MemoryInfoGetName and MemoryInfoGetMemType to determine it's GPU device memory.

So instead, add an API like this, that only works if OrtValue is a tensor and returns a pointer to the existing OrtMemoryInfo from the Tensor:

`ORT_API2_STATUS(GetTensorMemoryInfo, _In_ const OrtValue* value, _Out_ const OrtMemoryInfo** mem_info);`

Reply:
I have made the changes, please have a look

Reply:
Shouldnt GetTensorMemoryInfo be called GetValueMemoryInfo or GetMemoryInfo as it operates on OrtValue and not all OrtValues are Tensors?

Reply:
@RyanUnderhill suggested this name, as we internally access the API of tensor for this... Please suggest if this is final naming

Reply:
@smk2007 , the naming convention isn't to name what it operates on, rather what the Get() api returns.
it follows other api's such as GetTensorTypeAndShape(), GetTensorMutableData() etc. that takes in a Tensor OrtValue and returns data associated with the Tensor.

Reply:
@jywu-msft GetTensorTypeAndShape and GetTensorMutableData only operate on Tensors (even if the type is OrtValue) because type and shape and mutable data are APIs that are only callable on tensors. So those are fine.

However, if OrtMemoryInfo can be used in conjuntion with other non-Tensor OrtValues, then this API would be misnamed.

Can non-tensor values use this API?

Reply:
this particular api can only be used on tensors, so the naming is appropriate.
you bring up an interesting point, on whether there should be other api's for non-tensors
@ryanunderhill, @pranavsharma any thoughts?

Reply:
@jywu-msft Yeah, I knew it was well defined for Tensor types, but I wasn't sure this would apply to Maps/Sequences/Future types and if that would be useful for them.

If it's well defined and easy to add, it makes sense to do the general purpose one.

Reply:
The fact that maps/sequences are constructed using tensors internally is an implementation detail. There doesn't seem to be any use case asking for the location (mem info) of these types. Let's stick to tensor only for now.